mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -73,56 +73,187 @@ struct SqrtReciprocalOptimization : public OpRewritePattern<tosa::PowOp> {
+// %1 = tosa.select(%0, x, input)
+// to tosa.clamp{min = min, max = x}(input)
+// The first pattern occurs in decompositions of LeakyReLU/PReLU with an alpha
+// of zero
+struct SelectToClampOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+
+    auto geq = op.getPred().getDefiningOp<tosa::GreaterEqualOp>();
+    if (!geq) {
+      return rewriter.notifyMatchFailure(op,
+                                         "Predicate is not a GreaterEqualOp");
+    }
+
+    DenseElementsAttr geqIn2Attr;
+    if (!matchPattern(geq.getInput2(), m_Constant(&geqIn2Attr))) {
+      return rewriter.notifyMatchFailure(
+          op, "RHS of predicate GreaterEqualOp is not a constant");
+    }
+    auto isCompatibleSplat = [](DenseElementsAttr a,
+                                DenseElementsAttr b) -> bool {
+      if (!a.isSplat() || !b.isSplat()) {
+        return false;
+      }
+      if (llvm::isa<IntegerType>(a.getElementType())) {
+        return a.getSplatValue<APInt>() == b.getSplatValue<APInt>();
+      }
+      if (llvm::isa<FloatType>(a.getElementType())) {
+        return a.getSplatValue<APFloat>() == b.getSplatValue<APFloat>();
+      }
+      return false; // Only int and float types are supported
+    };
+
+    auto onFalse = op.getOnFalse();
+    auto onTrue = op.getOnTrue();
+    DenseElementsAttr onFalseAttr;
+    DenseElementsAttr onTrueAttr;
+
+    // Case one:
+    // %0 = tosa.greater_equal(input, cmp)
+    // %1 = tosa.select(%0, input, cmp)
+    // to tosa.clamp{min = cmp, max = max}(input)
+    // Predicate: geq.input2 == select.onFalse AND geq.input1 == select.onTrue
+    const bool isCaseOne =
+        matchPattern(onFalse, m_Constant(&onFalseAttr)) &&
+        isCompatibleSplat(onFalseAttr, geqIn2Attr) &&
+        onTrue.getDefiningOp() == geq.getInput1().getDefiningOp();
+
+    // Case two:
+    // %0 = tosa.greater_equal(input, cmp)
+    // %1 = tosa.select(%0, cmp, input)
+    // to tosa.clamp{min = input, max = cmp}(input)
+    // Predicate: geq.input2 == select.onTrue AND geq.input1 == select.onFalse
+    const bool isCaseTwo =
+        !isCaseOne && matchPattern(onTrue, m_Constant(&onTrueAttr)) &&
+        isCompatibleSplat(onTrueAttr, geqIn2Attr) &&
+        onFalse.getDefiningOp() == geq.getInput1().getDefiningOp();
+
+    if (!isCaseOne && !isCaseTwo) {
+      return rewriter.notifyMatchFailure(
+          op, "select does not match GEQ + select -> clamp pattern");
+    }
+
+    const auto inputElementType = geqIn2Attr.getElementType();
+    int64_t clampIntMin = std::numeric_limits<int64_t>::min();
+    int64_t clampIntMax = std::numeric_limits<int64_t>::max();
+    FloatAttr clampFloatMin;
+    FloatAttr clampFloatMax;
+    if (isa<IntegerType>(inputElementType)) {
+      auto splatValue = geqIn2Attr.getSplatValue<APInt>().trySExtValue();
```

Comment:
Our fork of LLVM allows for unsigned types (Tosa_Tensor def is not strictly following the spec) here as well could we guard against it? Or add the handling for unsigned. Either one.

Reply:
Added handling for unsigned with less than 63 bits